provider/cmd/pulumi-gen-eks/nodejs-templates/clusterMixins.ts

```diff
@@ -0,0 +1,160 @@
+import * as aws from '@pulumi/aws';
+import * as pulumi from '@pulumi/pulumi';
+import { NodeGroup } from './nodeGroup';
+import { Cluster, ClusterArgs } from './cluster';
+import { Provider } from '@pulumi/kubernetes';
+import { ClusterNodeGroupOptionsArgs, KubeconfigOptionsArgs, RoleMappingArgs } from './types/input';
+import { ResourceTransformArgs, ResourceTransformResult } from '@pulumi/pulumi';
+
+declare module "./cluster" {
+    interface Cluster {
+        provider: Provider;
+        createNodeGroup(name: string, args: ClusterNodeGroupOptionsArgs): NodeGroup;
+    }
+}
+
+export interface KubeconfigOptions extends KubeconfigOptionsArgs {}
+
+export interface RoleMapping extends RoleMappingArgs {}
+
+export interface ClusterOptions extends ClusterArgs {}
+
+Object.defineProperty(Cluster.prototype, '_provider', { writable: true, enumerable: false });
+
+Object.defineProperty(Cluster.prototype, 'provider', {
+    enumerable: true,
+    get(): Provider {
+        if (!this._provider) {
+            this._provider = new Provider(`${this.__name}-provider`, {
+                kubeconfig: this.kubeconfigJson,
+            }, { parent: this, aliases: [{ name: this.__name }]});
+        }
+        return this._provider;
+    }
+})
+
+Cluster.prototype.createNodeGroup = function(name: string, args: ClusterNodeGroupOptionsArgs): NodeGroup {
+    // const cluster = Cluster.prototype;
+    // const provider = await cluster.getAwsProvider();
+    // const awsProvider = provider.provider ? { aws: provider.provider } : undefined;
+
+    return new NodeGroup(
+        name,
+        {
+            ...args,
+            cluster: this.core,
+            nodeSecurityGroup: this.nodeSecurityGroup,
+            clusterIngressRule: this.eksClusterIngressRule,
+        },
+        {
+            // parent: this,
+            aliases: [{ parent: this }],
+            transforms: [
+                (args: ResourceTransformArgs): ResourceTransformResult => {
+                    return {
+                        props: args.props,
+                        opts: { ...args.opts, aliases: [{ parent: this }] },
+                    }
+                }
+            ],
+            providers: {
+                // TODO: need to get aws provider
+                // ...awsProvider,
```

Comment:
```suggestion
                ...(this.core.awsProvider ? { aws: this.core.awsProvider } : undefined),
```

Reply:
The problem is that `core.awsProvider` is an output type now which you can't pass to `providers`. I think the way we can solve this is to add an extra input to `createNodeGroup` which optionally takes this provider.

The `awsProvider` is not actually created inside `Cluster` it is the optional aws provider that the user can pass to `Cluster`. The user will just have to also pass the `awsProvider` to `createNodeGroup`.

Reply:
@flostadler I pushed up a new commit with the changes.

Reply:
Ah, yeah that makes sense.

We should also add this to the migration doc. This might lead to unexpected diffs for some users